torch/csrc/api/src/nn/modules/embedding.cpp

```diff
@@ -11,27 +12,186 @@
-  stream << "torch::nn::Embedding(count=" << options.count()
-         << ", dimension=" << options.dimension() << ")";
-}
+    void EmbeddingImpl::pretty_print(std::ostream& stream) const {
+      stream << "torch::nn::Embedding(num_embeddings=" << options.num_embeddings()
+             << ", embedding_dim=" << options.embedding_dim();
+      if (options.padding_idx() != c10::nullopt) {
+        stream << ", padding_idx=" << *options.padding_idx();
+      }
+      if (options.max_norm() != c10::nullopt) {
+        stream << ", max_norm=" << *options.max_norm();
+      }
+      if (options.norm_type() != 2) {
+        stream << ", norm_type=" << options.norm_type();
+      }
+      if (options.scale_grad_by_freq()) {
+        stream << ", scale_grad_by_freq=" << std::boolalpha << options.scale_grad_by_freq();
+      }
+      if (options.sparse()) {
+        stream << ", sparse=" << std::boolalpha << options.sparse();
+      }
+      stream << ")";
+    }
 
-Tensor EmbeddingImpl::forward(const Tensor& input) {
-  return torch::embedding(weight, /*indices=*/input);
-}
+    torch::Tensor EmbeddingImpl::forward(const Tensor& input) {
+      if (options.padding_idx() != c10::nullopt) {
+        if (*options.padding_idx() > 0) {
+          TORCH_CHECK(*options.padding_idx() < weight.size(0), "Padding_idx must be within num_embeddings");
+        }
+        else if (*options.padding_idx() < 0) {
+          TORCH_CHECK(*options.padding_idx() >= -weight.size(0), "Padding_idx must be within num_embedding");
+          options.padding_idx(weight.size(0) + *options.padding_idx());
+        }
+      } else {
+        options.padding_idx(-1);
+      }
+
+      if (options.max_norm() != c10::nullopt) {
+        torch::NoGradGuard no_grad;
+        torch::embedding_renorm_(weight, input.contiguous(), *options.max_norm(), options.norm_type());
+      }
+      return torch::embedding(weight, input.contiguous(), *options.padding_idx(), options.scale_grad_by_freq(), options.sparse());
+    }
+
+    Embedding Embedding::from_pretrained(const torch::Tensor& embeddings, c10::optional<EmbeddingOptions> options, bool freeze) {
+      TORCH_CHECK(embeddings.dim() == 2, "Embeddings parameter is expected to be 2-dimensional");
+      if (options != c10::nullopt) {
+        TORCH_CHECK((*options).num_embeddings() == embeddings.size(0), "Expects options.num_embeddings to be ", embeddings.size(0) , "but found ", (*options).num_embeddings());
+        TORCH_CHECK((*options).embedding_dim() == embeddings.size(1), "Expects options.embeddings_dim to be ", embeddings.size(1) , "but found ", (*options).embedding_dim());
+      } else {
+        options = EmbeddingOptions(embeddings.size(0), embeddings.size(1));
+      }
+      Embedding embedding = Embedding((*options)._weight(embeddings));
+      embedding->weight.set_requires_grad(!freeze);
+      return embedding;
+    }
+
+    EmbeddingBagImpl::EmbeddingBagImpl(const EmbeddingBagOptions& options) : options(options) {
```

Comment:
To address the clang-tidy error, we can add `// NOLINT(modernize-pass-by-value)` at the end of this line.

Reply:
Also don't forget `options` -> `options_` ;)